atorch/examples/ds_3d_parallel/design.md

```diff
@@ -0,0 +1,844 @@
+# Goal
+- Large-scale model (100B+ levels) training, in addition to the FSDP zero3 scheme, the community more uses hybrid parallel (data parallel, pipeline parallel, tensor parallel, 3D) scheme. The most widely used frameworks include Megatron-LM (tensor parallelism), deepspeed (pipeline parallelism), that form hybrid parallelism with deep optimization.
```

Comment:
For large-scale model training (with 100B+ levels), besides using FSDP/zero3 parallelism, 3D parallelism is widely used in deep learning community. 3D parallelism includes tensor parallel, pipeline parallel, and  data parallel.  Megatron-LM and DeepSpeed provide excellent 3D parallelism implementation which are popular among users. 
 
Megatron-LM offers a Col/Row parallel layer that users can incorporate into the model definition to achieve tensor parallelism. DeepSpeed's pipeline parallel feature requires converting the model into a sequential list of LayerSpec, making its usage complicated especially for non-sequential models.